packages/common/src/hooks/useDebounce/useDebounceCallback.test.tsx

```diff
@@ -0,0 +1,164 @@
+import { waitFor } from '@testing-library/dom';
+import { renderHook } from '@testing-library/react-hooks';
+import { useDebounceCallback } from './useDebounceCallback';
+
+describe('useDebounceCallback', () => {
+  it('returns a function which debounces consecutive calls', async () => {
+    // Setting up a simple value for incrementing and a function which is a closure
+    // over the value which will be incremented. The side effect makes this  somewhat flakey
+    // when creating tests, but allows us to access the real state (i) to compare against
+    // returned values from the callback to ensure they're the same.
+    let i = 0;
+    const func = () => {
+      i = i + 1;
+      return i;
+    };
+
+    const { result } = renderHook(() => {
+      const cb = useDebounceCallback(func, []);
+
+      return cb;
+    });
+
+    result.current();
+    result.current();
+    await result.current();
+
+    await waitFor(() => {
+      expect(i).toBe(1);
+    });
+  });
+
+  it('calls the function again correctly after a wait period has completed', async () => {
+    // Ensuring that after a wait period, we can just call the function again and the normal behaviour
+    // is resumed.
+    let i = 0;
+    const func = () => {
+      i = i + 1;
+      return i;
+    };
+
+    const { result } = renderHook(() => {
+      const cb = useDebounceCallback(func, [], 10);
+
+      return cb;
+    });
+
+    result.current();
+    result.current();
+    await result.current();
+    result.current();
+    await result.current();
+    await result.current();
+
+    await waitFor(() => {
+      expect(i).toBe(3);
+    });
+  });
+
+  it('returns a function which resolves to the final value', async () => {
+    let i = 0;
+    const func = () => {
+      i = i + 1;
+      return i;
+    };
+    // Here we set up a debounced callback and call it multiple times and compare that the
+    // resolved value from each invocation aligns with with what the 'state' of i "should be",
+    // which will mean that the resolved values are equal to the final state of the consecutive calls.
+    const { result } = renderHook(() => {
+      const cb = useDebounceCallback(func, [], 10);
+
+      return cb;
+    });
+
+    const val1 = result.current();
+    const val2 = result.current();
+    const val3 = result.current();
+
+    await waitFor(() => {
+      expect(val1).resolves.toBe(1);
```

Comment:
Ha!

Yes seems they were not quite doing what I thought..

I've resorted to using jest timer mocking. I had hoped to avoid it but timing related things are just a bit too annoying.

The test in particular was definitely a false positive. Turns out I am silly, and the the debounce isn't actually resolving all consecutive calls with the final value, rather only the last call is.